packages/runtime/container-runtime/src/opLifecycle/opSerialization.ts

```diff
@@ -0,0 +1,96 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	IFluidHandleContext,
+	type IFluidHandleInternal,
+} from "@fluidframework/core-interfaces/internal";
+import { assert } from "@fluidframework/core-utils/internal";
+import type { ISequencedDocumentMessage } from "@fluidframework/driver-definitions/internal";
+import { FluidSerializerBase } from "@fluidframework/runtime-utils/internal";
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+
+import type { OutboundContainerRuntimeMessage } from "../messageTypes.js";
+
+/**
+ * Takes an incoming runtime message (outer type "op"), JSON.parses the message's contents in place,
+ * if needed (old Loader does this for us).
+ * Only to be used for runtime messages. The contents here would be the virtualized payload for a batch of ops.
+ * @remarks - Serialization during submit happens via {@link OpSerializer.stringify}
+ * @param mutableMessage - op message received
+ */
+export function ensureContentsDeserialized(mutableMessage: ISequencedDocumentMessage): void {
+	// This should become unconditional once Loader LTS reaches 2.4 or later.
+	// There will be a long time of needing both cases, until LTS advances to that point.
+	if (typeof mutableMessage.contents === "string" && mutableMessage.contents !== "") {
+		mutableMessage.contents = JSON.parse(mutableMessage.contents);
+	}
+}
+
+// For passing to base class encode/stringify methods.
+const dummyBind: Partial<IFluidHandleInternal> = {
```

Comment:
This (plus the not-impl methods below) seems like a code smell - if we have to fake out the base class this much then maybe shouldn't use the base class (also would be nice to avoid class inheritance for its own merit).

Reply:
Trying to reuse code in this initial PR.  Although I'll note that this dummyBind doesn't work (the code REALLY wants it to be an actual handle), so maybe I'll just copy the code after all.

Reply:
This code definitely smells though, you're right 😁 would be something to clean up later in any case.  The whole Serializer types are really messed up w.r.t. handles and internal handles, I lost a lot of time fighting that losing battle last week (it's all legacy-alpha 😭)